File: python-language/scheduler.py
#!/usr/bin/env python3
"""
Weekly Shift Scheduler (Python)
Supports: morning, afternoon, evening
Employees provide preference ranking per day (list of shifts in priority order).
Implements constraints:
 - max 1 shift per day per employee
 - max 5 working days per week per employee
 - at least 2 employees per shift per day (randomly fill from eligible)
 - conflict resolution: try alternate same-day preference, then next days
"""

import random
from collections import defaultdict
import csv

DAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
SHIFTS = ["morning", "afternoon", "evening"]
MIN_PER_SHIFT = 2
MAX_DAYS_PER_EMP = 5

random.seed(42)

# ---------- Utility functions ----------
def init_schedule():
    return {day: {shift: [] for shift in SHIFTS} for day in DAYS}

def eligible_pool(employees, assignments, day, exclude_set):
    """Return employees eligible to be assigned on a given day"""
    pool = []
    for name in employees:
        if name in exclude_set:
            continue
        worked_days = len(assignments.get(name, set()))
        if worked_days >= MAX_DAYS_PER_EMP:
            continue
        pool.append(name)
    return pool

# ---------- Scheduling algorithm ----------
def schedule_week(employees):
    """
    employees: dict
      name -> {day: [pref1, pref2, pref3]}
    """
    schedule = init_schedule()
    assignments = defaultdict(set)  # name -> set(days worked)

    for day in DAYS:
        assigned_today = set()

        # Try to give each employee their top preference first
        for name, prefs in employees.items():
            if len(assignments[name]) >= MAX_DAYS_PER_EMP:
                continue
            top_pref = prefs.get(day, SHIFTS)[0]
            if name not in assigned_today and len(schedule[day][top_pref]) < MIN_PER_SHIFT:
                schedule[day][top_pref].append(name)
                assignments[name].add(day)
                assigned_today.add(name)

        # Fill shortages per shift
        for shift in SHIFTS:
            while len(schedule[day][shift]) < MIN_PER_SHIFT:
                pool = eligible_pool(employees, assignments, day, assigned_today)
                if not pool:
                    break
                # Prefer employees with this shift as 2nd/3rd preference
                ranked = []
                for emp in pool:
                    prefs = employees[emp].get(day, SHIFTS)
                    if shift in prefs:
                        ranked.append((prefs.index(shift), emp))
                    else:
                        ranked.append((99, emp))  # low priority if not in prefs
                ranked.sort()
                chosen = ranked[0][1]
                schedule[day][shift].append(chosen)
                assignments[chosen].add(day)
                assigned_today.add(chosen)

        # Conflict resolution: employees unassigned today
        for name, prefs in employees.items():
            if name in assigned_today:
                continue
            if len(assignments[name]) >= MAX_DAYS_PER_EMP:
                continue
            for alt_shift in prefs.get(day, SHIFTS):
                if name not in schedule[day][alt_shift]:
                    schedule[day][alt_shift].append(name)
                    assignments[name].add(day)
                    assigned_today.add(name)
                    break

    return schedule

# ---------- Output ----------
def pretty_print(schedule):
    print("\nFinal weekly schedule:\n" + "=" * 30)
    for day in DAYS:
        print(f"{day}:")
        for shift in SHIFTS:
            workers = schedule[day][shift]
            print(f"  {shift.title():<9}: {', '.join(workers) if workers else '(none)'}")
        print("-" * 30)

# ---------- Load Employees ----------
def load_employees_from_csv(filename):
    employees = {}
    with open(filename, newline="") as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            name = row["Name"]
            prefs = {}
            for i, day in enumerate(DAYS, start=1):
                prefs[day] = row[day].split(">")
            employees[name] = prefs
    return employees

# ---------- Main ----------
if __name__ == "__main__":
    employees = load_employees_from_csv(r"/workspaces/Assignment4-control-structure/python-language/employee.csv")
    schedule = schedule_week(employees)
    pretty_print(schedule)

File: javascript-langauge/scheduler.js
const fs = require('fs');

const DAYS = ["Mon","Tue","Wed","Thu","Fri","Sat","Sun"];
const SHIFTS = ["morning","afternoon","evening"];
const MIN_PER_SHIFT = 2;
const MAX_DAYS_PER_EMP = 5;

// Load employees from CSV with preference rankings
function loadEmployees(filename) {
    const data = fs.readFileSync(filename, 'utf-8');
    const lines = data.trim().split('\n');
    const header = lines.shift().split(',');
    let employees = {};
    for (const line of lines) {
        const row = line.split(',');
        const name = row[0];
        employees[name] = {};
        for (let i = 1; i < header.length; i++) {
            employees[name][header[i]] = row[i].split('>');
        }
    }
    return employees;
}

// Scheduler with priority + balancing
function scheduleWeek(employees) {
    const schedule = {};
    const assignments = {};
    for (let name in employees) assignments[name] = new Set();

    for (let day of DAYS) {
        schedule[day] = {morning: [], afternoon: [], evening: []};
        const assignedToday = new Set();

        // --- First pass: assign 1 per shift (based on preference) ---
        for (let shift of SHIFTS) {
            const candidates = Object.keys(employees)
                .filter(name => !assignedToday.has(name) && assignments[name].size < MAX_DAYS_PER_EMP)
                .map(name => {
                    const prefIndex = employees[name][day].indexOf(shift);
                    return {name, prefIndex};
                })
                .filter(c => c.prefIndex >= 0)
                .sort((a,b) => a.prefIndex - b.prefIndex);

            if (candidates.length > 0) {
                const c = candidates[0]; // best candidate
                schedule[day][shift].push(c.name);
                assignedToday.add(c.name);
                assignments[c.name].add(day);
            }
        }

        // --- Second pass: ensure 2 per shift (fill shortages) ---
        for (let shift of SHIFTS) {
            while (schedule[day][shift].length < MIN_PER_SHIFT) {
                // pool of people under 5 days and not already on this shift
                const pool = Object.keys(employees)
                    .filter(n => assignments[n].size < MAX_DAYS_PER_EMP && !schedule[day][shift].includes(n));

                if (pool.length === 0) break; // no more eligible
                const choice = pool[Math.floor(Math.random() * pool.length)];
                schedule[day][shift].push(choice);
                assignments[choice].add(day);
                assignedToday.add(choice);
            }
        }
    }

    return schedule;
}

module.exports = { loadEmployees, scheduleWeek };
